File: public/pedersen.js
// Pedersen / ElGamal primitives for the homomorphic tally protocol
// (Phase B / vote schema v3).
//
// Mirror of prover/pedersen.go. Pure BigInt — zero deps. Curve is
// BabyJubJub (twisted Edwards over BN254's scalar field Fr), the
// embedded curve used by gnark for in-circuit ECC over BN254 Groth16.
//
// Curve equation: a·x² + y² = 1 + d·x²·y² over Fr, with a = -1
// (gnark-crypto / iden3 reduced-parameter form).
//
// Encoding: 32-byte compressed (gnark-crypto / RFC 8032) — little-
// endian Y with the sign of X stored in the MSB of byte 31.
//
// H is the deterministic independent generator. The Go side derives
// it via try-and-increment (SHA-256 + cofactor-clearing); JS does not
// re-derive — it loads H as a fixed constant from pedersen_vectors.json
// and the parity test asserts byte-equality. H is a public parameter,
// so agreement on bytes is what matters, not algorithm portability.

// ---- Field & curve constants -----------------------------------------------

// BN254 scalar field — also the BabyJubJub coordinate field.
export const Fr = 21888242871839275222246405745257275088548364400416034343698204186575808495617n;

// Twisted-Edwards parameters (a, d) reduced into Fr.
const A = Fr - 1n;                    // a = -1 mod Fr
const D = 12181644023421730124874158521699555681764249180949974110617291017600649128846n;

// Prime-order subgroup order ℓ; cofactor = 8.
export const SUBGROUP_ORDER = 2736030358979909402780800718157159386076813972158567259200215660948447373041n;
export const COFACTOR = 8n;

// Canonical BabyJubJub base point (gnark-crypto / iden3 standard).
export const G = Object.freeze({
    x: 9671717474070082183213120605117400219616337014328744928644933853176787189663n,
    y: 16950150798460657717958625567821834550301663161624707787222815936182638968203n,
});

// Identity element (twisted Edwards): (0, 1).
export const ZERO = Object.freeze({ x: 0n, y: 1n });

const FR_HALF = (Fr - 1n) / 2n;
const FR_INV_EXP = Fr - 2n;          // by Fermat's little theorem

// Tonelli-Shanks parameters for sqrt mod Fr. BN254's Fr has 2-adicity
// 28, so simple p≡3 (mod 4) shortcut doesn't apply.
//   Fr − 1 = q · 2^s  with s = 28 and q odd
//   z      = 5  (a known quadratic non-residue mod Fr)
const TS_S = 28n;
const TS_Q = (Fr - 1n) >> TS_S;
const TS_Z = 5n;

// Compression flag bits (top bit of the *last* byte after little-endian
// serialization).
const FLAG_X_NEG = 0x80;
const FLAG_MASK  = 0x7f;

// ---- Field arithmetic ------------------------------------------------------

const mod = (a, m = Fr) => {
    const r = a % m;
    return r < 0n ? r + m : r;
};

function modPow(base, exp, m = Fr) {
    let r = 1n, b = mod(base, m), e = exp;
    while (e > 0n) {
        if (e & 1n) r = (r * b) % m;
        b = (b * b) % m;
        e >>= 1n;
    }
    return r;
}

const modInv = (a) => modPow(a, FR_INV_EXP);

// Tonelli-Shanks square-root mod Fr; returns null on non-residue.
function sqrtFr(n) {
    if (n === 0n) return 0n;
    // Euler's criterion: n^((Fr-1)/2) must be 1 for QR.
    if (modPow(n, FR_HALF) !== 1n) return null;

    let M = TS_S;
    let c = modPow(TS_Z, TS_Q);
    let t = modPow(n, TS_Q);
    let R = modPow(n, (TS_Q + 1n) >> 1n);
    while (true) {
        if (t === 1n) return R;
        let i = 0n;
        let temp = t;
        while (temp !== 1n) {
            temp = (temp * temp) % Fr;
            i += 1n;
            if (i === M) return null; // shouldn't happen if Euler passed
        }
        const b = modPow(c, 1n << (M - i - 1n));
        M = i;
        c = (b * b) % Fr;
        t = (t * c) % Fr;
        R = (R * b) % Fr;
    }
}

// fr_LexicographicallyLargest mirrors gnark's predicate: x is "negative"
// iff x > (Fr-1)/2.
const isLargest = (x) => x > FR_HALF;

// ---- Curve arithmetic (twisted-Edwards, unified addition) ------------------

export function isOnCurve(P) {
    if (!P) return false;
    const x2 = (P.x * P.x) % Fr;
    const y2 = (P.y * P.y) % Fr;
    const lhs = mod(A * x2 + y2);
    const rhs = mod(1n + D * x2 % Fr * y2);
    return lhs === rhs;
}

export const pointEq = (P, Q) => P.x === Q.x && P.y === Q.y;

export function pointNeg(P) {
    return { x: mod(-P.x), y: P.y };
}

// Unified twisted-Edwards addition (works for doubling too):
//   x3 = (x1·y2 + y1·x2) / (1 + d·x1·x2·y1·y2)
//   y3 = (y1·y2 − a·x1·x2) / (1 − d·x1·x2·y1·y2)
export function pointAdd(P, Q) {
    const x1y2 = P.x * Q.y % Fr;
    const y1x2 = P.y * Q.x % Fr;
    const y1y2 = P.y * Q.y % Fr;
    const x1x2 = P.x * Q.x % Fr;
    const dx1x2y1y2 = D * x1x2 % Fr * y1y2 % Fr;
    const denX = mod(1n + dx1x2y1y2);
    const denY = mod(1n - dx1x2y1y2);
    const x3 = mod((x1y2 + y1x2) * modInv(denX));
    const y3 = mod((y1y2 - A * x1x2) * modInv(denY));
    return { x: x3, y: y3 };
}

export const pointDouble = (P) => pointAdd(P, P);

export function scalarMul(k, P) {
    let n = mod(k, SUBGROUP_ORDER);
    if (n === 0n) return ZERO;
    let R = ZERO;
    let A_ = P;
    while (n > 0n) {
        if (n & 1n) R = pointAdd(R, A_);
        A_ = pointAdd(A_, A_);
        n >>= 1n;
    }
    return R;
}

export const scalarMulBase = (k) => scalarMul(k, G);

export function isInPrimeSubgroup(P) {
    return pointEq(scalarMul(SUBGROUP_ORDER, P), ZERO);
}

// ---- Encoding --------------------------------------------------------------

function bigintToLEBytes(x, len) {
    let v = x;
    const out = new Uint8Array(len);
    for (let i = 0; i < len; i++) {
        out[i] = Number(v & 0xffn);
        v >>= 8n;
    }
    if (v !== 0n) throw new Error(`bigint too large for ${len} LE bytes`);
    return out;
}

function leBytesToBigint(buf) {
    let v = 0n;
    for (let i = buf.length - 1; i >= 0; i--) {
        v = (v << 8n) | BigInt(buf[i]);
    }
    return v;
}

export function encodePoint(P) {
    const out = bigintToLEBytes(P.y, 32);
    if (isLargest(P.x)) out[31] |= FLAG_X_NEG;
    return out;
}

export function decodePoint(buf) {
    if (buf.length !== 32) throw new Error(`expected 32 bytes, got ${buf.length}`);
    const xNeg = (buf[31] & FLAG_X_NEG) !== 0;
    const yBuf = new Uint8Array(buf);
    yBuf[31] &= FLAG_MASK;
    const y = leBytesToBigint(yBuf);
    if (y >= Fr) throw new Error('y out of field');
    // x² = (y² − 1) / (d·y² − a)
    const ySq = y * y % Fr;
    const num = mod(ySq - 1n);
    const den = mod(D * ySq - A);
    if (den === 0n) throw new Error('decode: degenerate denominator');
    const xSq = mod(num * modInv(den));
    let x = sqrtFr(xSq);
    if (x === null) throw new Error('point not on curve (no sqrt)');
    if (isLargest(x) !== xNeg) x = mod(-x);
    const P = { x, y };
    if (!isOnCurve(P)) throw new Error('decoded point not on curve');
    return P;
}

export const encodePointHex = (P) =>
    Array.from(encodePoint(P), (b) => b.toString(16).padStart(2, '0')).join('');

export function decodePointHex(hex) {
    if (hex.length !== 64) throw new Error(`expected 64 hex chars, got ${hex.length}`);
    const buf = new Uint8Array(32);
    for (let i = 0; i < 32; i++) buf[i] = parseInt(hex.slice(i * 2, i * 2 + 2), 16);
    return decodePoint(buf);
}

// ---- ElGamal ---------------------------------------------------------------

export function encrypt(v, r, pk) {
    const vRed = mod(v, SUBGROUP_ORDER);
    const rRed = mod(r, SUBGROUP_ORDER);
    const A_ = scalarMulBase(rRed);
    const gv = scalarMulBase(vRed);
    const pkr = scalarMul(rRed, pk);
    const B_ = pointAdd(gv, pkr);
    return { A: A_, B: B_ };
}

export function aggregate(cts) {
    let A_ = ZERO, B_ = ZERO;
    for (const ct of cts) {
        A_ = pointAdd(A_, ct.A);
        B_ = pointAdd(B_, ct.B);
    }
    return { A: A_, B: B_ };
}

export class TallyExceedsRange extends Error {
    constructor() { super('pedersen: tally exceeds maxTally range'); }
}

export function decrypt(ct, sk, maxTally) {
    if (maxTally < 0) throw new Error('maxTally must be non-negative');
    const skA = scalarMul(sk, ct.A);
    const M = pointAdd(ct.B, pointNeg(skA));
    if (pointEq(M, ZERO)) return 0;
    let probe = ZERO;
    for (let t = 1; t <= maxTally; t++) {
        probe = pointAdd(probe, G);
        if (pointEq(probe, M)) return t;
    }
    throw new TallyExceedsRange();
}

// ---- H constant ------------------------------------------------------------
//
// Loaded from pedersen_vectors.json. Recompute on the Go side via
//   go test ./prover -run TestPedersenH -v

let H_CACHED = null;

export function setPedersenH(hex) {
    H_CACHED = decodePointHex(hex);
}

export function getPedersenH() {
    if (H_CACHED === null) {
        throw new Error('pedersen.js: H is not initialized — call setPedersenH(hex) first');
    }
    return H_CACHED;
}

File: prover/pedersen.go
// Package prover — Pedersen / ElGamal primitives for the homomorphic
// tally protocol (Phase B / vote schema v3).
//
// Curve: BabyJubJub — twisted Edwards curve embedded over BN254's
// scalar field Fr. Coordinates and arithmetic live natively in Fr,
// which is the constraint field of the BN254 Groth16 proof system —
// so in-circuit ElGamal scalar mults cost ~constant per bit instead
// of the ~50× overhead of non-native BN254 G1.
//
// Curve equation: a·x² + y² = 1 + d·x²·y², with a = -1 (gnark-crypto
// reduced parameters). Prime-order subgroup ℓ; cofactor = 8.
//
//   G = curve canonical base point (returned by gnark-crypto).
//   H = independent generator derived by hash-to-curve (try-and-increment
//       on SHA-256 of "bitwrap-h-generator-v1" + counter, post-multiplied
//       by the cofactor to land in the prime-order subgroup). Reusable
//       as a Pedersen commitment base. Not used by ElGamal itself but
//       defined for downstream constructions; both sides must agree on
//       the same byte sequence.
//
// ElGamal encoding (additively homomorphic, exponential, additive
// notation suitable for Edwards):
//
//	A = G · r
//	B = G · v + pk · r
//
// Aggregation is pointwise. Decryption:
//
//	M = B − sk · A   recovers G · v_total
//	v_total          via small-range search bounded by maxTally.
//
// Wire encoding: 32-byte compressed point in gnark-crypto / RFC 8032
// format — little-endian Y with the sign of X in the MSB of byte 31.
// JS parity is asserted by public/pedersen_parity_test.mjs against
// vectors emitted from prover/pedersen_test.go.

package prover

import (
	"crypto/sha256"
	"encoding/binary"
	"errors"
	"fmt"
	"math/big"
	"sync"

	"github.com/consensys/gnark-crypto/ecc/bn254/fr"
	tedwards "github.com/consensys/gnark-crypto/ecc/bn254/twistededwards"
)

// PedersenHDST is the domain separation tag used to derive H.
// Frozen — changing it produces an incompatible H.
const PedersenHDST = "bitwrap-h-generator-v1"

// Ciphertext is a single ElGamal ciphertext under the creator pk.
type Ciphertext struct {
	A tedwards.PointAffine
	B tedwards.PointAffine
}

// PedersenG returns the canonical BabyJubJub base point.
func PedersenG() tedwards.PointAffine {
	return tedwards.GetEdwardsCurve().Base
}

// PedersenH returns the deterministic independent generator H.
// Cached after first computation.
var (
	pedersenHOnce  sync.Once
	pedersenHValue tedwards.PointAffine
)

func PedersenH() tedwards.PointAffine {
	pedersenHOnce.Do(func() {
		pedersenHValue = deriveH(PedersenHDST)
	})
	return pedersenHValue
}

// deriveH performs a deterministic try-and-increment hash-to-curve:
//
//  1. counter = 0
//  2. seed  = SHA-256(dst || counter_be_4)
//  3. y     = seed mod Fr
//  4. solve x² = (y² − 1) / (d·y² − a) for x; if not a QR, ++counter.
//  5. pick the canonical x-root (lex-smallest), build (x, y).
//  6. multiply by cofactor=8 to land in the prime-order subgroup.
//  7. if result is identity (y=1, x=0), ++counter and retry.
//
// JS mirror does not reproduce this — H is loaded as a fixed constant
// from the parity-vector fixture (a single deterministic point that
// both sides must agree on byte-for-byte).
func deriveH(dst string) tedwards.PointAffine {
	curve := tedwards.GetEdwardsCurve()

	var aFr, dFr fr.Element
	aFr.Set(&curve.A)
	dFr.Set(&curve.D)

	for counter := uint32(0); counter < 1<<16; counter++ {
		var ctrBuf [4]byte
		binary.BigEndian.PutUint32(ctrBuf[:], counter)
		h := sha256.New()
		h.Write([]byte(dst))
		h.Write(ctrBuf[:])
		seed := h.Sum(nil)

		var y fr.Element
		y.SetBytes(seed) // reduces mod Fr

		// num = y² − 1, den = d·y² − a
		var ySq, num, den fr.Element
		ySq.Square(&y)
		num.Sub(&ySq, new(fr.Element).SetOne())
		den.Mul(&dFr, &ySq)
		den.Sub(&den, &aFr)
		if den.IsZero() {
			continue
		}
		var xSq fr.Element
		xSq.Div(&num, &den)

		// x = sqrt(xSq); Sqrt returns nil if non-residue.
		var x fr.Element
		if x.Sqrt(&xSq) == nil {
			continue
		}
		// canonicalize: pick lex-smallest x.
		if x.LexicographicallyLargest() {
			x.Neg(&x)
		}

		var p tedwards.PointAffine
		p.X.Set(&x)
		p.Y.Set(&y)
		if !p.IsOnCurve() {
			continue
		}
		// clear cofactor to land in the prime-order subgroup.
		var cof big.Int
		curve.Cofactor.BigInt(&cof)
		var cleared tedwards.PointAffine
		cleared.ScalarMultiplication(&p, &cof)
		if cleared.IsZero() {
			continue
		}
		return cleared
	}
	panic("pedersen: deriveH counter exhausted (DST is degenerate)")
}

// Encrypt returns the ElGamal ciphertext (A, B) for plaintext scalar v
// under public key pk and randomness r.
//
//	A = G · r
//	B = G · v + pk · r
//
// All scalars are reduced mod the BabyJubJub subgroup order before
// scalar mul. (gnark-crypto's PointAffine.ScalarMultiplication accepts
// any *big.Int but we canonicalize for parity with JS.)
func Encrypt(v, r *big.Int, pk *tedwards.PointAffine) Ciphertext {
	g := PedersenG()
	vRed := reduceSubgroup(v)
	rRed := reduceSubgroup(r)

	var A, B, gv, pkr tedwards.PointAffine
	A.ScalarMultiplication(&g, rRed)
	gv.ScalarMultiplication(&g, vRed)
	pkr.ScalarMultiplication(pk, rRed)
	B.Add(&gv, &pkr)
	return Ciphertext{A: A, B: B}
}

// Aggregate sums a list of ciphertexts component-wise. Empty input
// returns the identity (0, 1) on both A and B.
func Aggregate(cts []Ciphertext) Ciphertext {
	var A, B tedwards.PointAffine
	identityPoint(&A)
	identityPoint(&B)
	for i := range cts {
		A.Add(&A, &cts[i].A)
		B.Add(&B, &cts[i].B)
	}
	return Ciphertext{A: A, B: B}
}

// Decrypt recovers the aggregate plaintext from a ciphertext using
// secret key sk. maxTally bounds the small-range search; returns
// ErrTallyExceedsRange if no t in [0, maxTally] satisfies G·t = M.
//
//	M = B − sk · A
func Decrypt(ct Ciphertext, sk *big.Int, maxTally int) (int, error) {
	if maxTally < 0 {
		return 0, errors.New("pedersen: maxTally must be non-negative")
	}
	skRed := reduceSubgroup(sk)

	g := PedersenG()
	var skA, negSkA, M tedwards.PointAffine
	skA.ScalarMultiplication(&ct.A, skRed)
	negSkA.Neg(&skA)
	M.Add(&ct.B, &negSkA)

	var probe tedwards.PointAffine
	identityPoint(&probe)
	if probe.Equal(&M) {
		return 0, nil
	}
	for t := 1; t <= maxTally; t++ {
		probe.Add(&probe, &g)
		if probe.Equal(&M) {
			return t, nil
		}
	}
	return 0, ErrTallyExceedsRange
}

// ErrTallyExceedsRange — Decrypt found no t in [0, maxTally] matching M.
var ErrTallyExceedsRange = errors.New("pedersen: tally exceeds maxTally range")

// EncodePoint returns the 32-byte compressed encoding (gnark-crypto /
// RFC 8032 format).
func EncodePoint(p *tedwards.PointAffine) []byte {
	b := p.Bytes()
	return b[:]
}

// DecodePoint parses a 32-byte compressed point and verifies it lies
// on the curve. Does not enforce subgroup membership — callers that
// need it (e.g. accepting pk_creator from untrusted input) should call
// IsInPrimeSubgroup separately.
func DecodePoint(buf []byte) (tedwards.PointAffine, error) {
	var p tedwards.PointAffine
	if len(buf) != 32 {
		return p, fmt.Errorf("pedersen: expected 32 bytes, got %d", len(buf))
	}
	if _, err := p.SetBytes(buf); err != nil {
		return p, fmt.Errorf("pedersen: decode point: %w", err)
	}
	if !p.IsOnCurve() {
		return p, errors.New("pedersen: decoded point not on curve")
	}
	return p, nil
}

// IsInPrimeSubgroup checks that ℓ · p == identity, where ℓ is the
// prime-order subgroup order. Critical for any pk accepted from
// untrusted input — small-subgroup attacks otherwise leak sk bits.
func IsInPrimeSubgroup(p *tedwards.PointAffine) bool {
	curve := tedwards.GetEdwardsCurve()
	var probe tedwards.PointAffine
	probe.ScalarMultiplication(p, &curve.Order)
	return probe.IsZero()
}

// identityPoint sets p to the twisted-Edwards identity (0, 1).
func identityPoint(p *tedwards.PointAffine) {
	p.X.SetZero()
	p.Y.SetOne()
}

// reduceSubgroup normalizes a scalar to [0, ℓ).
func reduceSubgroup(x *big.Int) *big.Int {
	curve := tedwards.GetEdwardsCurve()
	return new(big.Int).Mod(x, &curve.Order)
}
